### PR TITLE
refactor(notes): extract pure logic and IO out of useNotesController

### PR DESCRIPTION
**Pull request by lpendrake**
## Summary

- **Problem**: `useNotesController.ts` (803 lines) contained business logic as plain function declarations inside the hook body. Those functions close over React state, are untestable without mounting React, and produced stale-closure hazards via nested `setTabs`/`setActiveTab` calls. The `open-note-by-path.test.ts` was the clearest symptom — it tested a hand-copied stub, not the real code.
- **Fix**: Extract all testable seams into their natural homes using existing conventions (no new folders invented). Replace correlated `tabs`/`activeTab` state with `useReducer`.
- **Rule**: Added a `CLAUDE.md` rule banning logic-inside-hooks going forward.

## What moved where

| Extracted | Destination | Reason |
|---|---|---|
| `parseNotePath` | `domain/open-note-by-path.ts` | Pure path manipulation, no IO |
| `resolveLinkById`, `resolveMarkdownHref` | `domain/link-resolution.ts` | Pure reads over the link index |
| `suggestLinks` | `shared/suggest-links.ts` | Editor feature, not notes-specific; deduplicates from `EventEditorModal` |
| `scanFolderContents` | `scan-folder.ts` (top-level) | IO-only; follows `editor-bindings.ts` / `data.ts` pattern |
| `tabs`/`activeTab` state | `tabsReducer` + `useReducer` | Eliminates nested-setState anti-pattern across 4 call sites; trivially unit-testable |
| `folderFiles`/`openFiles` mutations | `domain/file-ops.ts` | Pure dictionary transforms; replaces inline spread logic in every file-ops handler |

## Tests

- `__tests__/open-note-by-path.test.ts` (stub-based) — **deleted**
- `domain/__tests__/open-note-by-path.test.ts` — **new**, 5 cases for real `parseNotePath`
- `domain/__tests__/link-resolution.test.ts` — **new**, 11 cases across both helpers
- `domain/__tests__/tabs-reducer.test.ts` — **new**, 20 cases covering all 5 action types and edge cases
- `domain/__tests__/file-ops.test.ts` — **new**, 17 cases for all 7 pure functions
- `__tests__/scan-folder.test.ts` — **new**, 10 cases with `vi.mock` on `notesData`
- `shared/__tests__/suggest-links.test.ts` — **new**, 4 cases

1009 tests, 62 test files, all green. Lint clean.

## Test plan

- [ ] `npm test -- src/renderer/notes` — all notes tests pass
- [ ] `npm test` — full suite (1009) passes
- [ ] `npm run lint` — clean
- [ ] `npm run build` — builds without TypeScript errors

https://claude.ai/code/session_01JmGr7EcMvE1xU9XoxHn5Zn

### CODE REVIEW

File: CLAUDE.md [hunk 1]
```diff
@@ -49,6 +49,17 @@ Before opening the PR, spawn an opus advisor on the diff. The advisor must verif
...
+## No logic inside hooks or components
+
+Do not declare functions containing business logic inside a React hook or component body.
+Functions defined this way close over state and become untestable without mounting React.
+
+- Pure logic (no IO, no React) belongs in `domain/`
+- Non-React IO logic belongs as a named `.ts` file alongside `data.ts`
+- The hook body wires those up to React state — it does not contain the logic itself
+
+A test file that duplicates implementation logic in a local stub is the clearest sign this rule was broken.
```
Comment (lpendrake) [inline CLAUDE.md]: I feel like this was a one off brain fart, not a pattern worth calling out. No sane dev writes tests like this.

File: src/renderer/notes/AGENTS.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# `src/renderer/notes/` — Notes View (React)
+
+The folder/file browser, tabbed editor, and live markdown editor for
+notes (NPCs, locations, factions, plots, etc.). React, not vanilla DOM.
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Nothing in this project is vanilla dom, it's all react, no need to mention it.

File: src/renderer/notes/AGENTS.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# `src/renderer/notes/` — Notes View (React)
+
+The folder/file browser, tabbed editor, and live markdown editor for
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: The live markdown editor isn't in or for notes, it's for any and all markdown editing and viewing in any view. It's already been extracted and abstracted and is in use in peeks and event cards and event editing.

File: src/renderer/notes/AGENTS.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+```
+notes/
+  notes.tsx                   # view-level component
+  data.ts                     # thin wrapper over window.fsApi (IO boundary)
+  scan-folder.ts              # scanFolderContents — IO-only, no React
+  editor-bindings.ts          # pure config-builder functions
+  types.ts
+  domain/                     # pure functions, no React, no IO
+    slugify.ts
+    open-note-by-path.ts      # parseNotePath
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Is this duplicated? We have a search function which opens notes and events. Look for an opportunity to remove this code entirely and abstract whatever is already implemented enough for it to be reusable anywhere someone wants to cause a note to be opened.

File: src/renderer/notes/AGENTS.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    link-resolution.ts        # suggestLinks, resolveLinkById, resolveMarkdownHref
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Link suggestion is an editor feature, not a notes feature. Should it really be in a note specific directory? Does logic for this already exist? If it does can we abstract all the duplicate bits of logic out into the editor implementation?

File: src/renderer/notes/AGENTS.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  config logic. May import `data.ts` / `notesData`. No React. Tested in `__tests__/` with `vi.mock`.
+- **`hooks/`** — React state and effects. May import `domain/` and the top-level helpers.
+- **`components/`** — JSX only. Receive props; do not import `notesData` directly.
+
+## useNotesController ceiling
+
+The orchestrator is ~750 lines. The remaining length is file-ops handlers (`handleRename*`,
+`handleDelete*`, `handleMove`, `commitNew*`, `handleQuickAddCreate`) that each thread 5–10
+setState calls. Pulling these into a hook would mean passing every setter through a deps object —
+the abstraction would be larger than the code it hides.
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Is this code worth testing, though? The concern is that the code should work, and causing regressions should be hard because tests tell us what expected scenarios exist. If these functions you're describing are not testable logic, fine, but if logic crucial to behavior is being executed the. We should test it.

Maybe extraction is not the answer, but if not, then how do we test it? Have we followed patterns that make extraction hard, when we could have made it easy and low impact on the line count? Do we simply need a new kind of test that elts us exercise this code?

File: src/renderer/notes/AGENTS.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+**New pure logic goes in `domain/`. New IO-only logic goes as a named file alongside
+`scan-folder.ts`. Do not add logic to the orchestrator.** If a handler grows non-trivial
+computation, extract the computation to `domain/` and call it from the handler.
+
+## React conventions
+
+- One component per file. Filename matches the export.
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Kebab case file names only, across the codebase.

File: src/renderer/notes/AGENTS.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+- Hooks for state and effects.
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: What about them? Maybe it's my lack of understanding of react but this sentence doesn't say anything to me.

File: src/renderer/notes/AGENTS.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+- Components for JSX only.
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Jsx? No javascript allowed in this codebase, we use ts. Also I thought components har to go in tsx files, like you can't write that pseudo-html stuff anywhere else. Isn't this sentence both redundant and incorrect?

File: src/renderer/notes/AGENTS.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+- `domain/` for pure logic.
+- Top-level `.ts` files for non-React IO or config.
+
+## Don't
+
+- Don't call `window.fsApi` directly from React. Go through `data.ts` or a top-level helper.
+- Don't reimplement folder tree state — `useFolderTree` is the source of truth.
+- Don't put logic in the orchestrator that would be equally at home in `domain/` or a
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: It seems to me that the real pattern we're asking for is more along the lines making orchestrators which are light on logic, simply passing data around to various logical operators in the correct sequence.

These bundles of logic are extracted and ideally reusable but definitely testable in isolation.

File: src/renderer/notes/__tests__/scan-folder.test.ts [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  it('produces a note entry for a file matched in the index', async () => {
+    const index: LinkIndexEntry[] = [
+      { id: 'n1', path: 'notes/Lore/bob.md', title: 'Bob', type: 'note' },
+    ];
+    listFolder.mockResolvedValue([{ name: 'bob.md', isDirectory: false }]);
+    const result = await scanFolderContents(campaignPath, folder, index);
+    expect(result).toEqual([{ id: 'n1', path: 'bob.md', title: 'Bob', kind: 'note' }]);
+  });
+
+  it('produces an asset entry for an asset in the index', async () => {
```
Comment (lpendrake) [inline src/renderer/notes/__tests__/scan-folder.test.ts:41]: Could you try adding a test for paths that include a space?

I've had issues with folder names that include a space inside asset links, not sure where the bug is but we could check here with a quick test.

File: src/renderer/notes/domain/__tests__/link-resolution.test.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+describe('resolveLinkById', () => {
+  it('returns not-found for unknown id', () => {
+    expect(resolveLinkById(idx, 'nope')).toEqual({ kind: 'not-found' });
+  });
+
+  it('strips "timeline/" prefix for events', () => {
+    expect(resolveLinkById(idx, 'e1')).toEqual({ kind: 'event', filename: 'sess-01.md' });
+  });
+
+  it('splits notes path into folder + path (folder = first segment after "notes/")', () => {
```
Comment (lpendrake) [inline src/renderer/notes/domain/__tests__/link-resolution.test.ts:23]: Why would it do that?

**Review by lpendrake — COMMENTED** [review]
Let's call this a first pass.

File: src/renderer/notes/AGENTS.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    editor-tabs.tsx
+    breadcrumb-nav.tsx
+    meta-panel.tsx
+    note-context-menu.tsx
+    quick-add.tsx
+  __tests__/                  # cross-cutting tests
+  styles/
+```
+
+Note: link suggestion (`suggestLinks`) lives in `src/renderer/shared/suggest-links.ts` — it is
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: Why do we need to mention this?

File: src/renderer/notes/AGENTS.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+an editor feature shared with the event editor, not a notes-specific concern.
+
+## Layer rules
+
+- **`domain/`** — pure functions, no IO. Tested directly in `domain/__tests__/`.
+- **Top-level `.ts` files** (`scan-folder.ts`, `editor-bindings.ts`, `data.ts`) — non-React IO or
```
Comment (lpendrake) [inline src/renderer/notes/AGENTS.md]: What's the difference betwen this and domain logic? Both sound like they're pure logic files that are meant to be testable, neither is meant to have any react stuff in it.

File: src/renderer/notes/domain/__tests__/tabs-reducer.test.ts [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  });
+
+  it('is a no-op when the tab does not exist', () => {
+    const state: TabsState = { tabs: [lore], activeTab: lore };
+    const s = tabsReducer(state, { type: 'close', folder: 'NPCs', path: 'ghost.md' });
+    expect(s).toBe(state);
+  });
+});
+
+describe('close-folder', () => {
```
Comment (lpendrake) [inline src/renderer/notes/domain/__tests__/tabs-reducer.test.ts:85]: Close folder? I don't think that's a thing we can do... When is this invoked? Is iit in case we delete a whole folder with potentially many open documents in it?

File: src/renderer/timeline/event-editor/EventEditorModal.tsx [hunk 16]
```diff
@@ -1,7 +1,7 @@
 import { useState, useEffect, useRef, useCallback, useMemo } from 'react';
 import type { EditorView } from '@codemirror/view';
 import { MarkdownEditor, FormatToolbar } from '../../shared/markdown-editor';
-import type { WikiLinkSuggestion } from '../../shared/markdown-editor';
```
Comment (lpendrake) [inline src/renderer/timeline/event-editor/EventEditorModal.tsx:4]: If there's already code for suggesting links, why are we adding new code?

Most likely you should have deleted the suggest link code and reused the already abstracted code.

File: src/renderer/notes/domain/__tests__/tabs-reducer.test.ts (hunk 15 above)
Comment (lpendrake) [reply src/renderer/notes/domain/__tests__/tabs-reducer.test.ts:85]: Correct — `close-folder` fires inside `handleDeleteFolder` to atomically close all tabs from the deleted folder. The name is from the tabs reducer's perspective ("close all tabs that belong to this folder"), not a UI action. Fixed the AGENTS.md note about this; will also rename the action to `delete-folder-tabs` if that reads more clearly — let me know.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/renderer/timeline/event-editor/EventEditorModal.tsx (hunk 16 above)
Comment (lpendrake) [reply src/renderer/timeline/event-editor/EventEditorModal.tsx:4]: This PR removed code, not added it. `EventEditorModal` had a 13-line duplicate `suggestLinks` implemented inline as a `useCallback`. That duplicate was deleted and replaced with this single import from the shared module. Net change to this file: −13 lines, +1 import.

---
_Generated by [Claude Code](https://claude.ai/code)_
